Support MPS Backend also on iOS < 16 (#9089)

f-meloni opened the pull request:
Summary: Pull Request resolved: https://github.com/pytorch/executorch/pull/9089

Differential Revision: D70795041

The MPS backend currently does not compile for iOS < 16.
I have added some conditional code to make it usable anyway

f-meloni commented:
cc @DenisVieriu97

f-meloni commented:
@pytorchbot label "topic: not user facing"

shoumikhin commented:
@DenisVieriu97 please take a look!

shoumikhin commented:
@DenisVieriu97 shall we just merge it?

cccclai commented on backends/apple/mps/runtime/operations/IndexingOps.mm:
```diff
@@ -206,24 +206,27 @@
-    Debug, "%s %d: %d",
-    __FUNCTION__, graphNode->input1_id(), graphNode->output_id()
-  );
+  if (@available(iOS 16, macOS 13, *)) {
```
Should we return an error for the else statement

f-meloni replied:
If the if is not true, then we do not do `_idToMPSGraphTensor[graphNode->output_id()] = ...` which I think would then not set the tensor value and that would fail the prediction, but only if this is used.
Do you think we should return `Error::notSupported` or something like that?

cccclai replied:
Yeah, just we can handle it in a non-fatal way.

f-meloni replied:
@cccclai Is the latest version OK?
Thank you for the review :)

cccclai commented on backends/apple/mps/runtime/operations/IndexingOps.mm:
```diff
@@ -206,25 +206,30 @@
+
+    return Error::Ok;
+  } else {
+    return Error::NotSupported;
```
Thanks, one minor thing - can you add an error log here like
```
ET_LOG(Error, "...")
```

f-meloni replied:
Added, and I also changed the requirements to iOS 15.4 and MacOS 12.3

Commit:
Add MPS Backend (#9095)

Summary:
Pull Request resolved: https://github.com/pytorch/executorch/pull/9095

Pull Request resolved: https://github.com/pytorch/executorch/pull/9089

Reviewed By: cccclai

Differential Revision: D70795041

DenisVieriu97 commented on backends/apple/mps/runtime/MPSDevice.mm:
```diff
@@ -22,11 +22,11 @@ static inline MTLLanguageVersion getMetalLanguageVersion(const id<MTLDevice>& de
-#if defined(__MAC_13_0)
-  if (macOS13Plus) {
-    languageVersion = MTLLanguageVersion3_0;
+  if (@available(iOS 16, macOS 13, *)) {
```
Thank you for the change

DenisVieriu97 commented on backends/apple/mps/runtime/operations/IndexingOps.mm:
```diff
@@ -206,25 +206,32 @@
-    Debug, "%s %d: %d",
-    __FUNCTION__, graphNode->input1_id(), graphNode->output_id()
-  );
+  if (@available(iOS 15.4, macOS 12.3, *)) {
```
good catch

DenisVieriu97 reviewed (approved):
Looks good - thanks @f-meloni for the changes
